python-api/gstswitch/server.py

```diff
@@ -22,6 +24,107 @@
 TOOLS_DIR = '/'.join(os.getcwd().split('/')[:-1]) + '/tools/'
 
 
+class ProcessMonitor(object):
+    """Runs a Command in a Background-Thread and monitors it's output
+
+    Can block until the Command prints a certain string and log the full
+    output into a file
+    """
+
+    def __init__(self, cmd, logtarget=sys.stderr):
+        self.log = logging.getLogger('server-output-monitor')
+
+        # After calling start(), _proc contains a subprocess.Popen instance
+        self._proc = None
+
+        # Command to run (Array passed to subprocess.Popen)
+        self._cmd = cmd
+
+        # Logfile to write to
+        self._logtarget = logtarget
+
+        # Internal Buffer to search for matched when wait_for_output is called
+        self._buffer = ""
+
+    def start(self):
+        """Starts up the process specified in the constructor in a thread.
+        Blocks until the process is started and has a pid assigned, but not
+        longer then 0.5s.
+        """
+
+        assert self._proc is None
+        self.log.debug("starting subprocess")
+
+        self._proc = subprocess.Popen(
+            self._cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            bufsize=0,
+            shell=False)
```

Comment:
I tried different approaches over the last weeks including reader threads and pexpect and none of them was as stable as this.

I can see the blocking problem when the server fills up the kernel buffer though. Is there any reason to prefer the file approach over the pty one? The pty idea seems to be cleaner to me,